libyul/AsmParser.cpp

```diff
@@ -78,6 +64,65 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
 	return nullptr;
 }
 
+langutil::Token Parser::advance()
+{
+	auto const token = ParserBase::advance();
+	updateLocation();
+	return token;
+}
+
+void Parser::updateLocation()
+{
+	if (m_scanner->currentCommentLiteral().empty())
+		return;
+
+	// parse doxygen style comment with a sequence of key/value pairs
+	// TODO: [ ] docflag!
+	// TODO: [x] join regexes into one, hard search for @src
+	// TODO: [x] catch exception on stoi (e.g. too biig number?)
+	// TODO: [ ] updateLocationEndFrom(): pull top if-condition out of it
+
+	static std::regex const lineRE = std::regex(
+		R"~~~(.*@src\s+(\d+):(-1|\d+):(-1|\d+).*)~~~",
+		std::regex_constants::ECMAScript | std::regex_constants::optimize
+	);
+
+	string_view text = m_scanner->currentCommentLiteral();
+	std::cmatch cm;
+	if (!std::regex_search(text.data(), text.data() + text.size(), cm, lineRE))
+		return;
+
+	try
+	{
+		solAssert(cm.size() == 4, "");
+		auto const sourceIndex = static_cast<unsigned>(stoul(cm[1].str()));
```

Comment:
does the static_cast truncate?

Reply:
stoul uses strtoul which sets errno ERANGE which will be translated to `std::out_of_range{}` It will return an `unsigned long` and we expect just an `unsigned int`, so it will be truncated if within range of ulong. I think I'll guard against that case.